### PR TITLE
Introduce CredentialsUtil for simplified credentials management.

### PR DESCRIPTION
**Pull request by boaks**
Make certificates optional.

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

File: demo-apps/cf-secure/src/main/java/org/eclipse/californium/examples/CredentialsUtil.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+		if (psk && config.getIncompleteConfig().getPskStore() == null) {
+			// Pre-shared secret keys
+			InMemoryPskStore pskStore = new InMemoryPskStore();
+			pskStore.setKey(PSK_IDENTITY, PSK_SECRET);
+			config.setPskStore(pskStore);
+		}
+		int x509 = indexOf(Mode.X509, modes);
+		int rpk = indexOf(Mode.RPK, modes);
+
+		if (0 <= x509 || 0 <= rpk) {
```
Comment (sbernard31) [inline demo-apps/cf-secure/src/main/java/org/eclipse/californium/examples/CredentialsUtil.java]: I do not understand why you always put constant in first.
I know this is logically  the same but it seems more readable to put the variable in first. 
I mean you want check if `x509` greater or equals to `0`, not if `0` is lesser or equals than `x509` ...
Almost all the code was written like that and if I remember well Kai already ask you for that.

File: demo-apps/cf-secure/src/main/java/org/eclipse/californium/examples/CredentialsUtil.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+	 * 
+	 * @param mode mode to check for
+	 * @param modes modes to check
+	 * @return {@code true}, if mode is contained in modes, or {@code false}, if
+	 *         not
+	 * @throws NullPointerException if one of the provided parameters are
+	 *             {@code null}
+	 */
+	public static boolean contains(Mode mode, Mode... modes) {
+		return 0 <= indexOf(mode, modes);
```
Comment (sbernard31) [inline demo-apps/cf-secure/src/main/java/org/eclipse/californium/examples/CredentialsUtil.java]: why not using Arrays.asList().containt()

File: demo-apps/cf-secure/src/main/java/org/eclipse/californium/examples/CredentialsUtil.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+	 * 
+	 * @param config DTLS configuration builder. May be already initialized with
+	 *            PskStore.
+	 * @param certificateAlias alias for certificate to load as credentials.
+	 * @param modes list of supported mode. If a RPK is in the list before X509,
+	 *            the RPK is setup as preferred
+	 */
+	public static void setupCredentials(DtlsConnectorConfig.Builder config, String certificateAlias, Mode... modes) {
+
+		boolean psk = 0 <= indexOf(Mode.PSK, modes);
```
Comment (sbernard31) [inline demo-apps/cf-secure/src/main/java/org/eclipse/californium/examples/CredentialsUtil.java]: why not using Arrays.asList().indexOf

File: demo-apps/cf-secure/src/main/java/org/eclipse/californium/examples/CredentialsUtil.java (hunk 1 above)
Comment (boaks) [reply demo-apps/cf-secure/src/main/java/org/eclipse/californium/examples/CredentialsUtil.java]: That's a habit from coding C with MISRA :-).

Comment (sophokles73) [reply demo-apps/cf-secure/src/main/java/org/eclipse/californium/examples/CredentialsUtil.java]: I have to agree with @sbernard31, I would also very much prefer if you could drop this habit of yours from the *good old days* (at least while working on Cf) ;-)

File: demo-apps/cf-secure/src/main/java/org/eclipse/californium/examples/CredentialsUtil.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+	/**
+	 * Parse arguments to modes.
+	 * 
+	 * @param args arguments
+	 * @return array of modes. if the arguments are empty,
+	 *         {@link #DEFAULT_MODES} is returned.
+	 */
+	public static List<Mode> parse(String[] args) {
+		int len = args.length;
+		if (0 == len) {
```
Comment (sbernard31) [inline demo-apps/cf-secure/src/main/java/org/eclipse/californium/examples/CredentialsUtil.java]: Still remains a MISRA's memory :) (this is one is less confusing)

I think the use of  `len` variable is not really needed. 
Unless you eventually reuse it below in `new ArrayList<>(len)`, personnaly I prefer to use args.length in the first `if`

**Pull request merged**